tests/unit/test_matrix.cu

```diff
@@ -2,64 +2,399 @@
 #include <vector>
 
 #include "../gtest.h"
+#include "common.hpp"
 
 #include <math.hpp>
 #include <matrix.hpp>
 #include <backends/fvm_gpu.hpp>
+#include <backends/fvm_multicore.hpp>
 #include <memory/memory.hpp>
 #include <util/span.hpp>
 
-using matrix_type = nest::mc::matrix<nest::mc::gpu::backend>;
-using index_type = matrix_type::size_type;
+#include <backends/gpu_kernels.hpp>
 
-TEST(matrix, solve_gpu)
+using namespace nest::mc;
+
+using gpu::impl::npos;
+using util::make_span;
+using util::assign_from;
+using memory::on_gpu;
+using memory::on_host;
+
+using testing::seq_almost_eq;
+
+using std::begin;
+using std::end;
+
+template <typename T>
+bool operator ==(const std::vector<T>& l, const std::vector<T>& r) {
+    if (l.size()!=r.size()) {
+        return false;
+    }
+
+    for (auto i=0u; i<l.size(); ++i) {
+        if (l[i]!=r[i]) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+template <typename T>
+bool operator !=(const std::vector<T>& l, const std::vector<T>& r) {
+    return !(l==r);
+}
+
+// will test the interleaving and reverse_interleaving operations for the
+// set of matrices defined by sizes and starts.
+// Applies the interleave to the vector in values, and checks this against
+// a reference result generated using a host side reference implementation.
+// Then the interleave result is reverse_interleaved, and the result is
+// compared to the original input.
+//
+// This is implemented in a separate function to facilitate testing on a
+// broad range of BlockWidth and LoadWidth compile time parameters.
+template <typename T, typename I, int BlockWidth, int LoadWidth>
+void test_interleave(
+        std::vector<I> sizes,
+        std::vector<I> starts,
+        std::vector<T> values,
+        int padded_size)
 {
-    using namespace nest::mc;
+    auto num_mtx = sizes.size();
+
+    auto in  = on_gpu(memory::make_const_view(values));
+    auto sizes_d = on_gpu(memory::make_const_view(sizes));
+    auto starts_d = on_gpu(memory::make_const_view(starts));
+
+    int packed_size = padded_size * BlockWidth * gpu::impl::block_count(num_mtx, BlockWidth);
+
+    // forward will hold the result of the interleave operation on the GPU
+    auto forward = memory::device_vector<T>(packed_size, npos<T>());
+
+    // find the reference interleaved values using host side implementation
+    auto baseline = gpu::interleave_host(values, sizes, starts, BlockWidth, num_mtx, padded_size);
+
+    // find the interleaved values on gpu
+    gpu::interleave<T, I, BlockWidth, LoadWidth>(in.data(), forward.data(), sizes_d.data(), starts_d.data(), padded_size, num_mtx);
+
+    std::vector<T> result_f = assign_from(on_host(forward));
+    std::vector<T> expected = gpu::interleave_host(values, sizes, starts, BlockWidth, num_mtx, padded_size);
+    const auto forward_success = (result_f==expected);
+    EXPECT_TRUE(forward_success);
```

Comment:
`test_interleave` should return a `::testing::AssertionResult` that's tested at the caller's end, or else the callers should use `SCOPED_TRACE` to indicate from where and with what this test is being called. Otherwise, it is hard to determine which invocation of the `test_interleave` call failed.